### PR TITLE
Add lie algebra for matrix SE(n)

### PR DESCRIPTION
**Pull request by nguigs**
This addresses #879 and will allow to use #878 on the Special Euclidean group

### CODE REVIEW

File: geomstats/geometry/skew_symmetric_matrices.py [hunk 1]
```diff
@@ -52,6 +52,22 @@ def belongs(self, mat, atol=TOLERANCE):
         """
         return Matrices(self.n, self.n).is_skew_symmetric(mat=mat, atol=atol)
 
+    @staticmethod
+    def projection(mat):
+        """Make matrix skew-symmetric, by averaging with minus its transpose.
```
Comment (ninamiolane) [inline geomstats/geometry/skew_symmetric_matrices.py:57]: NIT: Maybe here an equation would be clearer than the sentence in English

File: geomstats/geometry/lie_algebra.py [hunk 2]
```diff
@@ -135,17 +136,66 @@ def matrix_representation(self, basis_representation):
 
         return gs.einsum("...i,ijk ->...jk", basis_representation, self.basis)
 
+    def reshape_metric_matrix(self, metric_matrix):
```
Comment (ninamiolane) [inline geomstats/geometry/lie_algebra.py:139]: Would this be better as a function within the module `matrices.py` ?

File: geomstats/geometry/lie_algebra.py [hunk 3]
```diff
@@ -135,17 +136,66 @@ def matrix_representation(self, basis_representation):
...
+            Symmetric matrix.
+        """
+        if Matrices.is_diagonal(metric_matrix):
+            metric_coeffs = gs.diagonal(metric_matrix)
+            metric_mat = gs.abs(
+                self.matrix_representation(metric_coeffs))
+            return metric_mat
+        raise ValueError('This is only possible for a diagonal matrix')
+
     def orthonormal_basis(self, metric_matrix):
```
Comment (ninamiolane) [inline geomstats/geometry/lie_algebra.py:166]: would this be better within the module `riemannian_metric.py`? Having metric-related functions within the Lie algera module could be confusing?

File: geomstats/geometry/special_euclidean.py [hunk 4]
```diff
@@ -9,7 +9,9 @@
 from geomstats.geometry.euclidean import Euclidean
 from geomstats.geometry.general_linear import GeneralLinear
 from geomstats.geometry.invariant_metric import InvariantMetric
+from geomstats.geometry.lie_algebra import MatrixLieAlgebra
```
Comment (ninamiolane) [inline geomstats/geometry/special_euclidean.py:12]: Should we distinguish MatrixLieGroup and LieGroup then?

File: geomstats/geometry/special_euclidean.py [hunk 5]
```diff
@@ -36,13 +38,46 @@
                         - 1. / 362880.]
 
 
+def embed(rotation, translation, output_shape, constant=1.):
+    r"""Embed rotation, translation couples into n+1 square matrices.
```
Comment (ninamiolane) [inline geomstats/geometry/special_euclidean.py:42]: I think this is called the homogeneous matrix representation. Maybe rename the function and add a link?

File: geomstats/geometry/special_euclidean.py [hunk 6]
```diff
@@ -36,13 +38,46 @@
...
+    output_shape : tuple of int
+        Desired output shape. This is need for vectorization.
+    constant : float
+        Constant to use at the last line and column of the square matrix.
+        Optional, default: 1
+
+    Returns
+    -------
+    mat: array-like, shape=[..., n + 1, n + 1]
+        Squre Matrix of size n + 1. It can represent an element of the
```
Comment (ninamiolane) [inline geomstats/geometry/special_euclidean.py:62]: Squre -> Square

Comment (ninamiolane) [reply geomstats/geometry/special_euclidean.py:62]: Not sure I understand this docstring? special euclidean group of the Lie algebra?

File: tests/test_skew_symmetric_matrices.py [hunk 7]
```diff
@@ -61,7 +61,7 @@ def test_bch_up_to_fourth_order_works(self):
             )
             self.assertAllClose(expected, result)
 
-    @geomstats.tests.np_only
+    @geomstats.tests.np_and_pytorch_only
```
Comment (ninamiolane) [inline tests/test_skew_symmetric_matrices.py:64]: 🎉 always exciting!

**Review by ninamiolane — APPROVED** [review]
Very nice!

**Comment by ninamiolane** [comment]
On a high-level, in terms of the structure and inheritance of classes:

- Since there is now MatrixLieAlgebra, do we want to have MatrixLieGroup? It could help both for consistency in our naming, but also acknowledging that all of our groups are matrix lie groups so far.

- Should we factorize Euclidean and Matrices into the same (more abstract) class, as in `pymanopt`:

Screen Shot 2020-11-17 at 9 07 00 AM

File: geomstats/geometry/lie_algebra.py (hunk 2 above)
Comment (nguigs) [reply geomstats/geometry/lie_algebra.py:139]: It needs access to `self.matrix_representation` and `self.basis`

Comment (nguigs) [reply geomstats/geometry/lie_algebra.py:139]: So both should go there? But the MatrixLieAlgbera class doesn't inherit the Matrices class

File: geomstats/geometry/special_euclidean.py (hunk 4 above)
Comment (nguigs) [reply geomstats/geometry/special_euclidean.py:12]: I think LieGroup with the point_type argument works ok.

File: geomstats/geometry/special_euclidean.py (hunk 5 above)
Comment (nguigs) [reply geomstats/geometry/special_euclidean.py:42]: True

File: geomstats/geometry/special_euclidean.py (hunk 6 above)
Comment (nguigs) [reply geomstats/geometry/special_euclidean.py:62]: Nit: it can represent an element of SE(3) or its Lie aglebra

**Comment by nguigs** [comment]
>     * Since there is now MatrixLieAlgebra, do we want to have MatrixLieGroup? It could help both for consistency in our naming, but also acknowledging that all of our groups are matrix lie groups so far.

MatrixLieAlgebra was here before (since the first hackathon).

>     * Should we factorize Euclidean and Matrices into the same (more abstract) class, as in `pymanopt`:

This makes sense, however, the point of the Matrices class was to gather all the current matrix operations in a well vectorized way, more as an extention of the backend than as mathematical space. In Euclidean we won't have the transpose, mul, to_symmetric etc... which are the useful methods of the current Matrices class

File: geomstats/geometry/lie_algebra.py (hunk 3 above)
Comment (nguigs) [reply geomstats/geometry/lie_algebra.py:166]: Yes, but we need a basis!

Comment (ninamiolane) [reply geomstats/geometry/lie_algebra.py:166]: The basis could be passed in argument?

Comment (nguigs) [reply geomstats/geometry/lie_algebra.py]: Yes makes sense actually, and to also move reshape metric matrix

**Pull request merged**